cpp/lib/Authenticator.cpp

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright (c) 2022 Snowflake Computing, Inc. All rights reserved.
+ */
+
+#include <string>
+#include <algorithm>
+#include <cctype>
+#include <chrono>
+#include <regex>
+#include <functional>
+
+#include "Authenticator.hpp"
+#include "../logger/SFLogger.hpp"
+#include "error.h"
+
+#include <openssl/pem.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+
+#include <fstream>
+
+#include "snowflake/Simba_CRTFunctionSafe.h"
+
+#ifdef __APPLE__
+#include <CoreFoundation/CFBundle.h>
+#include <ApplicationServices/ApplicationServices.h>
+#endif
+
+#ifdef _WIN32
+#include <Shellapi.h>
+#define strcasecmp _stricmp
+#endif
+
+#define JWT_THROW(msg)                              \
+{                                                   \
+  throw Snowflake::Client::Jwt::JwtException(msg);  \
+}
+
+// wrapper functions for C
+extern "C" {
+
+  AuthenticatorType getAuthenticatorType(const char* authenticator)
+  {
+    if ((!authenticator) || (strlen(authenticator) == 0) ||
+        (strcasecmp(authenticator, SF_AUTHENTICATOR_DEFAULT) == 0))
+    {
+      return AUTH_SNOWFLAKE;
+    }
+    if (strcasecmp(authenticator, SF_AUTHENTICATOR_JWT) == 0)
+    {
+      return AUTH_JWT;
+    }
+
+    return AUTH_UNSUPPORTED;
+  }
+
+  SF_STATUS STDCALL auth_initialize(SF_CONNECT * conn)
+  {
+    if (!conn)
+    {
+      return SF_STATUS_ERROR_NULL_POINTER;
+    }
+
+    AuthenticatorType auth_type = getAuthenticatorType(conn->authenticator);
+    try
+    {
+      if (AUTH_JWT == auth_type)
+      {
+        conn->auth_object = static_cast<Snowflake::Client::IAuthenticator*>(
+                              new Snowflake::Client::AuthenticatorJWT(conn));
+      }
+    }
+    catch (...)
+    {
+      SET_SNOWFLAKE_ERROR(&conn->error, SF_STATUS_ERROR_GENERAL,
+                          "authenticator initialization failed",
+                          SF_SQLSTATE_GENERAL_ERROR);
+      return SF_STATUS_ERROR_GENERAL;
+    }
+
+    return SF_STATUS_SUCCESS;
+  }
+
+  int64 auth_get_renew_timeout(SF_CONNECT * conn)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return 0;
+    }
+
+    try
+    {
+      return static_cast<Snowflake::Client::IAuthenticator*>(conn->auth_object)
+               ->getAuthRenewTimeout();
+    }
+    catch (...)
+    {
+      return 0;
+    }
+  }
+
+  SF_STATUS STDCALL auth_authenticate(SF_CONNECT * conn)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return SF_STATUS_SUCCESS;
```

Comment:
Should we be returning success in this case?

Reply:
I don't think !conn would happen but !conn->auth_object is a normal case that extra authenticator is not needed, like user/password or oauth.